### PR TITLE
Full Body IK

### PR DESCRIPTION
**Pull request by HexaField**
## Summary
Full Body IK Implementation

- [x] pre-cache height & other necessary measurements based on bone structure in rig component upon avatar load
  - [x] h: head to foot, a: head to hips, b: hips to knee, c: knee to feet
- [x] write a simple solver for the head/hip/knee/feet offsets
  - [x] calculate internal angle (A) based on cosine rule
  - [x] calculate opposite angle base on sine rule against right angle
  - [x] get z, y from resulant triangle
- [x] do two bone solve from from hips to feet using knee as swing/twist
- [x] apply angle to hips to orient torso
- [x] rotate feet to make them flat to the ground
- [x] ensure translating & rotating the avatar does not break (only been using the developer test suite examples which have the test avatar at the original)
- [x] ensure existing arm and head look IK work as intended (may need to update matrices)
- [x] test with headsets & ensure networking works correctly

Other Fixes
- [x] debug helper & gizmo state fixes used in testing
- [x] improve spectate mode api

## References

closes https://github.com/XRFoundation/XREngine/issues/7486

## Checklist
- [ ] If this PR is still a WIP, convert to a draft
- [ ] [ensure all checks pass](https://github.com/XRFoundation/XREngine/wiki/Testing-&-Contributing)
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewer

## QA Steps

_List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos._

### CODE REVIEW

File: packages/engine/src/renderer/EngineRendererState.ts [hunk 1]
```diff
@@ -62,53 +63,59 @@ function setUsePostProcessing(usePostProcessing) {
 export class EngineRendererReceptor {
   static setQualityLevel(action: typeof EngineRendererAction.setQualityLevel.matches._TYPE) {
     const s = getState(EngineRendererState)
-    s.merge({ qualityLevel: action.qualityLevel })
+    s.qualityLevel.set(action.qualityLevel)
```
Comment (mrhegemon) [inline packages/engine/src/renderer/EngineRendererState.ts:66]: I assume this is an unrelated fix?

File: packages/engine/src/avatar/components/AvatarAnimationComponent.ts [hunk 2]
```diff
@@ -43,13 +45,28 @@ export const AvatarRigComponent = defineComponent({
...
-      helper: null as SkeletonHelper | null
+      helper: null as SkeletonHelper | null,
+      torsoLength: 0,
+      upperLegLength: 0,
+      lowerLegLength: 0,
+      footHeight: 0
     }
   },
 
   onSet: (entity, component, json) => {
```
Comment (mrhegemon) [inline packages/engine/src/avatar/components/AvatarAnimationComponent.ts:56]: Can we put some more comments here?

**Review by mrhegemon — APPROVED** [review]
Hip solver is so cool.

File: packages/engine/src/renderer/EngineRendererState.ts (hunk 1 above)
Comment (HexaField) [reply packages/engine/src/renderer/EngineRendererState.ts:66]: yes, fixed the debuggers to make it easier to debug stuff

**Review by speigg — APPROVED** [review]
